README.md
dcepucp
=======

[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active) [![Travis-CI Build Status](https://travis-ci.org/sebschub/dcepucp.svg?branch=master)](https://travis-ci.org/sebschub/dcepucp) [![codecov](https://codecov.io/gh/sebschub/dcepucp/branch/master/graph/badge.svg)](https://codecov.io/gh/sebschub/dcepucp)

Overview
--------

dcepucp is an R package that creates a NetCDF file from urban canopy
parameters as required by DCEP
([Schubert et al. 2012](http://dx.doi.org/10.1007/s10546-012-9728-3))
implemented in [COSMO-CLM](http://www.clm-community.eu/) version 5. In
particular, it provides the functions

  * `ugrid()` to define an urban grid including COSMO-CLM parameters
    and DCEP parameters
  * `upar()` to define the complete set of urban parameters including
    building and street properties as well as grid information
  * `upar2nc()` to save an `upar` object in a NetCDF file

Furthermore, the package includes an exemplary data set of Berlin,
Germany used
in
[Schubert and Grossman-Clarke 2013](http://doi.org/10.1127/0941-2948/2013/0393). The
fields are called `berlin_*` (see below).


Installation
------------

Install the package from github using the devtools package with
``` r
# install.packages("devtools")
devtools::install_github("sebschub/dcepucp")
```
If you receive the error `Installation failed: error in running
command`, try setting `options(unzip = "unzip")` before running `install_github`.


Usage
-----

The following shows the usage of dcepucp using the example of the
Berlin data set:
``` r
library(dcepucp)
# generature urban grid (equivalent to berlin_grid already defined by
# the package
berlin_grid <- ugrid(pollat = 37.483, pollon = -166.6,
                     startlat_tot = -0.882, startlon_tot = -0.882,
                     ie_tot = 195, je_tot = 195,
                     dlat = 0.009, dlon = 0.009,
                     ke_uhl = 13,
                     hhl_uhl = c(0, 5, 10, 15, 20, 25, 30, 35, 40, 50, 60, 80, 100, 140),
                     n_uclass = 1,
                     n_udir = 4, angle_udir = c(-45., 0., 45., 90.))

# define urban parameters
ucp <- upar(berlin_grid, fr_urb = berlin_fr_urb, 
            fr_uclass = berlin_fr_uclass, fr_udir = berlin_fr_udir,
            fr_roof = berlin_fr_roof, w_street = berlin_w_street,
            w_build = berlin_w_build)

# save urban parameters in ucp.nc
upar2nc("ucp.nc", ucp)
```


NEWS.md
# dcepucp 0.2.99

* Add derivation of urban parameters from Local Climate Zones information

# dcepucp 0.2.7

* Corrected metadata of fr_udir in NetCDF output

# dcepucp 0.2.6

* Added `polgam` argument to `ugrid`

# dcepucp 0.2.5

* Changed street direction convention in example data: angle increases
  clockwise with 0 pointing north

# dcepucp 0.2.4

* Added a `NEWS.md` file to track changes to the package.
* Added `grid_mapping` and `coordinates` attributes to fields in the netCDF output



tests/testthat.R
library(testthat)
library(dcepucp)

test_check("dcepucp")


tests/testthat/test_lcz.R
context("LCZ")

test_that("validation", {
  expect_silent(validate_lcz(lcz_default))

  lcz <- lcz_default

  lcz[[3]]$fr_build <- 200
  lcz[[2]]$h2w <- 20000

  expect_warning(validate_lcz(lcz),
                 regexp = "h2w in class \"Compact midrise\" is out of range.")
  expect_warning(validate_lcz(lcz),
                 regexp = "fr_build in class \"Compact low-rise\" is out of range.")


  expect_silent(set_lcz_parameter(lcz_default, "Compact high-rise", h2w = 3))
  lcz2 <- set_lcz_parameter(lcz_default, "Compact high-rise", h2w = 3)
  expect_s3_class(lcz2, "lcz")

})


tests/testthat/test_objects.R
context("Objects")

bgrid <- ugrid(pollat = 37.483, pollon = -166.6,
               startlat_tot = -0.882, startlon_tot = -0.882,
               ie_tot = 195, je_tot = 195,
               dlat = 0.009, dlon = 0.009,
               ke_uhl = 13,
               hhl_uhl = c(0, 5, 10, 15, 20, 25, 30, 35, 40, 50, 60, 80, 100, 140),
               n_uclass = 1,
               n_udir = 4, angle_udir = c(-45., 0., 45., 90.))


test_that("ugrid", {
  expect_equal(bgrid, berlin_grid)
  expect_is(bgrid, "ugrid")
})

brlat <- rotated_latitude(bgrid)
brlon <- rotated_longitude(bgrid)

brlat_test <-
c(-0.882, -0.873, -0.864, -0.855, -0.846, -0.837, -0.828, -0.819,
  -0.81, -0.801, -0.792, -0.783, -0.774, -0.765, -0.756, -0.747,
  -0.738, -0.729, -0.72, -0.711, -0.702, -0.693, -0.684, -0.675,
  -0.666, -0.657, -0.648, -0.639, -0.63, -0.621, -0.612, -0.603,
  -0.594, -0.585, -0.576, -0.567, -0.558, -0.549, -0.54, -0.531,
  -0.522, -0.513, -0.504, -0.495, -0.486, -0.477, -0.468, -0.459,
  -0.45, -0.441, -0.432, -0.423, -0.414, -0.405, -0.396, -0.387,
  -0.378, -0.369, -0.36, -0.351, -0.342, -0.333, -0.324, -0.315,
  -0.306, -0.297, -0.288, -0.279, -0.27, -0.261, -0.252, -0.243,
  -0.234, -0.225, -0.216, -0.207, -0.198, -0.189, -0.18, -0.171,
  -0.162, -0.153, -0.144, -0.135, -0.126, -0.117, -0.108, -0.099,
  -0.09, -0.081, -0.072, -0.063, -0.054, -0.045, -0.036, -0.027,
  -0.018, -0.009, 0, 0.009, 0.018, 0.027, 0.036, 0.045, 0.054,
  0.063, 0.072, 0.081, 0.09, 0.099, 0.108, 0.117, 0.126, 0.135,
  0.144, 0.153, 0.162, 0.171, 0.18, 0.189, 0.198, 0.207, 0.216,
  0.225, 0.234, 0.243, 0.252, 0.261, 0.27, 0.279, 0.288, 0.297,
  0.306, 0.315, 0.324, 0.333, 0.342, 0.351, 0.36, 0.369, 0.378,
  0.387, 0.396, 0.405, 0.414, 0.423, 0.432, 0.441, 0.45, 0.459,
  0.468, 0.477, 0.486, 0.495, 0.504, 0.513, 0.522, 0.531, 0.54,
  0.549, 0.558, 0.567, 0.576, 0.585, 0.594, 0.603, 0.612, 0.621,
  0.63, 0.639, 0.648, 0.657, 0.666, 0.675, 0.684, 0.693, 0.702,
  0.711, 0.72, 0.729, 0.738, 0.747, 0.756, 0.765, 0.774, 0.783,
  0.792, 0.801, 0.81, 0.819, 0.828, 0.837, 0.846, 0.855, 0.864)

brlon_test <-
c(-0.882, -0.873, -0.864, -0.855, -0.846, -0.837, -0.828, -0.819,
  -0.81, -0.801, -0.792, -0.783, -0.774, -0.765, -0.756, -0.747,
  -0.738, -0.729, -0.72, -0.711, -0.702, -0.693, -0.684, -0.675,
  -0.666, -0.657, -0.648, -0.639, -0.63, -0.621, -0.612, -0.603,
  -0.594, -0.585, -0.576, -0.567, -0.558, -0.549, -0.54, -0.531,
  -0.522, -0.513, -0.504, -0.495, -0.486, -0.477, -0.468, -0.459,
  -0.45, -0.441, -0.432, -0.423, -0.414, -0.405, -0.396, -0.387,
  -0.378, -0.369, -0.36, -0.351, -0.342, -0.333, -0.324, -0.315,
  -0.306, -0.297, -0.288, -0.279, -0.27, -0.261, -0.252, -0.243,
  -0.234, -0.225, -0.216, -0.207, -0.198, -0.189, -0.18, -0.171,
  -0.162, -0.153, -0.144, -0.135, -0.126, -0.117, -0.108, -0.099,
  -0.09, -0.081, -0.072, -0.063, -0.054, -0.045, -0.036, -0.027,
  -0.018, -0.009, 0, 0.009, 0.018, 0.027, 0.036, 0.045, 0.054,
  0.063, 0.072, 0.081, 0.09, 0.099, 0.108, 0.117, 0.126, 0.135,
  0.144, 0.153, 0.162, 0.171, 0.18, 0.189, 0.198, 0.207, 0.216,
  0.225, 0.234, 0.243, 0.252, 0.261, 0.27, 0.279, 0.288, 0.297,
  0.306, 0.315, 0.324, 0.333, 0.342, 0.351, 0.36, 0.369, 0.378,
  0.387, 0.396, 0.405, 0.414, 0.423, 0.432, 0.441, 0.45, 0.459,
  0.468, 0.477, 0.486, 0.495, 0.504, 0.513, 0.522, 0.531, 0.54,
  0.549, 0.558, 0.567, 0.576, 0.585, 0.594, 0.603, 0.612, 0.621,
  0.63, 0.639, 0.648, 0.657, 0.666, 0.675, 0.684, 0.693, 0.702,
  0.711, 0.72, 0.729, 0.738, 0.747, 0.756, 0.765, 0.774, 0.783,
  0.792, 0.801, 0.81, 0.819, 0.828, 0.837, 0.846, 0.855, 0.864)

test_that("rotated longitude and latitude", {
  expect_equal(brlat, brlat_test)
  expect_equal(brlon, brlon_test)
})


blat <- latitude(bgrid)
blon <- longitude(bgrid)

test_that("geographical longitude and latitude", {
  expect_equal(dim(blat), c(bgrid$ie_tot, bgrid$je_tot))
  expect_equal(dim(blon), c(bgrid$ie_tot, bgrid$je_tot))
  expect_lt(max(blat), 53.4)
  expect_gt(min(blat), 51.6)
  expect_lt(max(blon), 14.9)
  expect_gt(min(blon), 11.9)
})

ucp <- upar(bgrid, fr_urb = berlin_fr_urb,
            fr_uclass = berlin_fr_uclass, fr_udir = berlin_fr_udir,
            fr_roof = berlin_fr_roof, w_street = berlin_w_street,
            w_build = berlin_w_build)

test_that("upar", {
  expect_length(ucp, 7)
  expect_is(ucp, "upar")
})


R/lcz.R
validate_lcz_parameter <- function(name_class, name_property, value_property) {
  if ((!is.na(lcz_ranges[[name_class]][[name_property]]$max) &
       value_property > lcz_ranges[[name_class]][[name_property]]$max) |
      (!is.na(lcz_ranges[[name_class]][[name_property]]$min) &
       value_property < lcz_ranges[[name_class]][[name_property]]$min) ) {

    warning(paste0(name_property, " in class \"", name_class, "\" is out of range."))
  }
}

#' Validate a parameter set
#'
#' A Local Climate Zones parameter set is validated against parameter ranges.
#'
#' @param lcz The LCZ parameter set.
#'
#'
#' @export
#'
#' @examples
#' validate_lcz(lcz_default)
validate_lcz <- function(lcz) {

  stopifnot(class(lcz) == "lcz")

  for (i_class in seq_along(lcz_ranges)) {
    name_class <- names(lcz_ranges)[i_class]

    # all properties of each class
    for (i_property in seq_along(lcz_ranges[[name_class]])) {

      name_property <- names(lcz_ranges[[name_class]])[i_property]
      validate_lcz_parameter(name_class, name_property, lcz[[name_class]][[name_property]])

    }
  }

}

#' Show LCZ class parameter, ranges and validate settings
#'
#' Print the Local Climate Zones parameter, its ranges and validate it.
#'
#' @param x LCZ parameter of class "lcz".
#' @param ... Not used.
#'
#' @export
#'
#' @examples
#' print(lcz_default)
print.lcz <- function(x, ...) {

  stopifnot(class(x) == "lcz")

  # all LCZ classes
  for (i_class in seq_along(lcz_ranges)) {

    cat(paste0("LCZ ", sprintf("%2d", i_class), ": "))
    name_class <- names(lcz_ranges)[i_class]
    cat(name_class)
    cat("\n")

    # all properties of each class
    for (i_property in seq_along(lcz_ranges[[i_class]])) {

      name_property <- names(lcz_ranges[[i_class]])[i_property]
      cat(paste0("  ", name_property, ": ", x[[name_class]][[name_property]], " "))

      # print ranges
      if (is.na(lcz_ranges[[i_class]][[i_property]]$min)) {
        cat(paste0("(< ", lcz_ranges[[i_class]][[i_property]]$max, ")"))
      } else if (is.na(lcz_ranges[[i_class]][[i_property]]$max)) {
        cat(paste0("(> ", lcz_ranges[[i_class]][[i_property]]$min, ")"))
      } else {
        cat(paste0("(", lcz_ranges[[i_class]][[i_property]]$min,
                   " - ", lcz_ranges[[i_class]][[i_property]]$max, ")"))
      }
      cat("\n")
    }
  }
  validate_lcz(x)
}


#' Set LCZ parameter
#'
#' Modify one or several parameters of one Local Climate Zones. The values are
#' checked for consistency with the valid LCZ range.
#'
#' @param lcz LCZ parameter, object of class \code{lcz}.
#' @param lcz_name Number or character string of the LCZ.
#' @param h2w The building-height-to-street-width ratio.
#' @param fr_build Fraction of buildings.
#' @param fr_street Fraction of street surfaces.
#' @param height Average building height.
#'
#' @export
#'
#' @examples
#' lcz <- set_lcz_parameter(lcz_default, "Compact high-rise", h2w = 3)
#' lcz
set_lcz_parameter <- function(lcz, lcz_name, h2w = NULL, fr_build = NULL, fr_street = NULL,
                             height = NULL) {

  stopifnot(class(lcz) == "lcz")

  # list of all arguments expect first two
  properties <- as.list(environment())[c(-1, -2)]

  # loop over all properties in the list; update if non-null
  for (i_property in seq_along(properties)) {
    if (!is.null(properties[[i_property]])) {

      name_property <- names(properties)[i_property]
      value_property <- properties[[i_property]]

      # check ranges if ranges are defined (non-NA)
      validate_lcz_parameter(name_class = lcz_name, name_property, value_property)
      lcz[[lcz_name]][name_property] <- value_property
    }
  }
  lcz
}


upar_lcz_dominant_class <- function(grid, lcz, dominant_class, fr_urb) {

  fr_uclass <- array(1./grid$n_uclass, dim = c(grid$ie_tot, grid$je_tot, grid$n_uclass))
  fr_udir <- array(1./grid$n_udir, dim = c(grid$ie_tot, grid$je_tot, grid$n_udir, grid$n_uclass))
  fr_roof <- array(0., dim = c(grid$ie_tot, grid$je_tot, grid$ke_uhl + 1, grid$n_udir, grid$n_uclass))

  w_street <- array(NA_real_, dim = c(grid$ie_tot, grid$je_tot, grid$n_udir, grid$n_uclass))
  w_build <- array(NA_real_, dim = c(grid$ie_tot, grid$je_tot, grid$n_udir, grid$n_uclass))

  h2w <- array(NA_real_, dim = c(grid$ie_tot, grid$je_tot))
  fr_build <- array(NA_real_, dim = c(grid$ie_tot, grid$je_tot))
  fr_street <- array(NA_real_, dim = c(grid$ie_tot, grid$je_tot))
  height <- array(NA_real_, dim = c(grid$ie_tot, grid$je_tot))

  for (i in 1:grid$ie_tot) {
    for (j in 1:grid$je_tot) {

      if (!is.na(dominant_class[i,j])) {

        h2w[i, j] <- lcz[[ dominant_class[i,j] ]]$h2w
        fr_build[i, j] <- lcz[[ dominant_class[i,j] ]]$fr_build
        fr_street[i, j] <- lcz[[ dominant_class[i,j] ]]$fr_street
        height[i, j] <- lcz[[ dominant_class[i,j] ]]$height

        if (lcz[[ dominant_class[i,j] ]]$height > max(grid$hhl_uhl))
          stop("Grid too low for dominant class \"", dominant_class[i,j], "\".")

        height_index <- which.min(abs(grid$hhl_uhl - lcz[[ dominant_class[i,j] ]]$height))

        fr_roof[i, j, height_index, , 1] <- 1.
      }
    }
  }

  for (id in 1:grid$n_udir) {
    w_street[ , , id, 1] <- height / h2w
    w_build[ , , id, 1] <- fr_build / fr_street * w_street[ , , id, 1]
  }

  upar(grid, fr_urb, fr_uclass, fr_udir, fr_roof, w_street, w_build)

}


#' Generate DCEP paramters from LCZ maps
#'
#' The parameters can be derived either from a map of the dominant class or a
#' list of fractions of local climate zones.
#'
#' @param grid Grid specifications, object of class \code{ugrid}.
#' @param lcz LCZ parameters, object of class \code{lcz}.
#' @param dominant_class Map of dominant class of each grid cell.
#' @param class_fractions List of maps, each describing the fraction of urban
#'   local climate zones.
#'
#' @return An object of class \code{upar}.
#' @details Either an argument for \code{dominent_class} or for
#'   \code{class_fraction} has to be supplied. Currently, only one urban class
#'   is supported.
#'
#' @export
#'
#' @examples
#' berlin_grid <- ugrid(pollat = 37.483, pollon = -166.6,
#' startlat_tot = -0.882, startlon_tot = -0.882,
#' ie_tot = 4, je_tot = 5,
#' dlat = 0.009, dlon = 0.009,
#' ke_uhl = 3,
#' hhl_uhl = c(0, 5, 10, 15),
#' n_uclass = 1,
#' n_udir = 4, angle_udir = c(-45., 0., 45., 90.))
#'
#' up <- upar_lcz(berlin_grid, lcz_default, class_fractions = list("Open low-rise" = array(1, c(4,5))))
upar_lcz <- function(grid, lcz, dominant_class = NULL, class_fractions = NULL) {

  stopifnot(class(grid) == "ugrid")
  stopifnot(grid$n_uclass == 1)
  stopifnot(xor(is.null(dominant_class), is.null(class_fractions)))

  fr_urb <- array(0., dim = c(grid$ie_tot, grid$je_tot))

  if (!is.null(class_fractions)) {
    stopifnot(class(class_fractions) == "list")
    # check if dimensions are ok
    for (i_lcz in seq_along(class_fractions)) {
    #  stopifnot_fraction_or_na(lcz_maps[[i_lcz]])
      stopifnot(dim(class_fractions[[i_lcz]]) == c(grid$ie_tot, grid$je_tot))
    }
    class_fractions <- simplify2array(class_fractions)

    stopifnot(all( apply(class_fractions, c(1,2), mean, na.rm = TRUE) <= 1.))

    lcz_names <- dimnames(class_fractions)[[3]]

    dominant_class <- array(NA_integer_, dim = c(grid$ie_tot, grid$je_tot))

    for (i in 1:grid$ie_tot) {
      for (j in 1:grid$je_tot) {

        # fr_urb based on average of all classes
        for (lcz_i in seq_along(lcz_names)) {
          if (!is.na(class_fractions[i,j, lcz_i])) {
            fr_urb[i,j] <- fr_urb[i,j] +
              class_fractions[i,j, lcz_i]*(lcz[[lcz_names[lcz_i]]]$fr_build + lcz[[lcz_names[lcz_i]]]$fr_street)
          }
        }
        # derive dominant class
        dominant_class[i, j] <- which(names(lcz) == lcz_names[which.max(class_fractions[i, j, ])])
      }
    }
  } else if (!is.null(dominant_class)) {

    for (i in 1:grid$ie_tot) {
      for (j in 1:grid$je_tot) {
        # fr_urb based on given class
        if (!is.na(dominant_class[i,j])) {
          fr_urb[i,j] <- lcz[[ dominant_class[i,j] ]]$fr_build +
                         lcz[[ dominant_class[i,j] ]]$fr_street
        }
      }
    }
  }

  upar_lcz_dominant_class(grid = grid, lcz = lcz,
                          dominant_class = dominant_class, fr_urb = fr_urb)

}


R/netcdf_output.R
#' Define NetCDF variable with standard attributes
#'
#' The variable is defined and attributes used in COSMO-CLM are added.
#'
#' @param nc NetCDF file object.
#' @param varname String of variable name.
#' @param dimensions Vector of dimension names as strings.
#' @param standard_name String of standard name.
#' @param long_name String of long name.
#' @param units String of units.
#' @param add_grid_mapping Add \code{grid_mapping} and \code{coordinates}
#'   attributes?
#'
#' @importFrom RNetCDF var.def.nc att.put.nc
#' @keywords internal
#' @examples
#' nc <- RNetCDF::create.nc("rot_lon.nc")
#' RNetCDF::dim.def.nc(nc, dimname = "rlon", dimlength = 100)
#' dcepucp:::nc_def_var(nc, varname = "rlon", dimensions = "rlon",
#'                      standard_name = "grid_longitude",
#'                      long_name = "rotated longitude",
#'                      units = "degrees")
#' RNetCDF::var.put.nc(nc, variable ="rlon", data = 1:100)
#' RNetCDF::close.nc(nc)
nc_def_var <- function(nc,
                       varname, dimensions, standard_name, long_name, units,
                       add_grid_mapping = TRUE) {
  var.def.nc(nc, varname = varname, vartype = "NC_FLOAT", dimensions = dimensions)
  att.put.nc(nc, variable = varname, name = "standard_name", type = "NC_CHAR",
             standard_name)
  att.put.nc(nc, variable = varname, name = "long_name", type = "NC_CHAR",
             long_name)
  att.put.nc(nc, variable = varname, name = "units", type = "NC_CHAR",
             units)
  att.put.nc(nc, variable = varname, name = "_FillValue", type = "NC_FLOAT",
             -1.e+20)
  if (add_grid_mapping) {
    att.put.nc(nc, variable = varname, name = "grid_mapping", type = "NC_CHAR",
               "rotated_pole")
    att.put.nc(nc, variable = varname, name = "coordinates", type = "NC_CHAR",
               "lon lat")
  }

}

#' Save urban parameters into NetCDF
#'
#' An \code{upar} object is stored in a NetCDF file. This file includes all
#' variables of the input object with a standard set of NetCDF attributes.
#' Information about the rotated pole are included as well.
#'
#' @param file String of output NetCDF file.
#' @param ucp \code{upar} object.
#'
#' @export
#' @importFrom RNetCDF create.nc close.nc dim.def.nc var.def.nc att.put.nc var.put.nc
#' @examples
#' ucps <- upar(berlin_grid, fr_urb = berlin_fr_urb,
#'              fr_uclass = berlin_fr_uclass, fr_udir = berlin_fr_udir,
#'              fr_roof = berlin_fr_roof, w_street = berlin_w_street,
#'              w_build = berlin_w_build)
#' upar2nc("ucp.nc", ucps)
upar2nc <- function(file, ucp) {

  stopifnot(is.character(file))
  stopifnot(class(ucp) == "upar")

  nc <- create.nc(file)

  # definition
  dim.def.nc(nc, dimname = "rlon", dimlength = ucp$grid$ie_tot)
  nc_def_var(nc, varname = "rlon", dimensions = "rlon",
             standard_name = "grid_longitude",
             long_name = "rotated longitude",
             units = "degrees",
             add_grid_mapping = FALSE)

  dim.def.nc(nc, dimname = "rlat", dimlength = ucp$grid$je_tot)
  nc_def_var(nc, varname = "rlat", dimensions = "rlat",
             standard_name = "grid_latitude",
             long_name = "rotated latitude",
             units = "degrees",
             add_grid_mapping = FALSE)

  dim.def.nc(nc, dimname = "uclass", dimlength = ucp$grid$n_uclass)

  dim.def.nc(nc, dimname = "udir", dimlength = ucp$grid$n_udir)
  nc_def_var(nc, varname = "udir", dimensions = "udir",
             standard_name = "street_direction",
             long_name = "street direction",
             units = "degrees",
             add_grid_mapping = FALSE)

  dim.def.nc(nc, dimname = "uheight1", dimlength = ucp$grid$ke_uhl + 1)
  nc_def_var(nc, varname = "uheight1", dimensions = "uheight1",
             standard_name = "urban_height_half_level",
             long_name = "height above surface for half levels",
             units = "m",
             add_grid_mapping = FALSE)

  nc_def_var(nc, varname = "lon", dimensions = c("rlon", "rlat"),
             standard_name = "longitude",
             long_name = "longitude",
             units = "degrees_east",
             add_grid_mapping = FALSE)
  nc_def_var(nc, varname = "lat", dimensions = c("rlon", "rlat"),
             standard_name = "latitude",
             long_name = "latitude",
             units = "degrees_north",
             add_grid_mapping = FALSE)

  var.def.nc(nc, varname = "rotated_pole", vartype = "NC_CHAR", dimensions = NA)
  att.put.nc(nc, variable = "rotated_pole", name = "long_name",
             type = "NC_CHAR",
             "coordinates of the rotated North Pole")
  att.put.nc(nc, variable = "rotated_pole", name = "grid_mapping_name",
             type = "NC_CHAR",
             "rotated_latitude_longitude")
  att.put.nc(nc, variable = "rotated_pole", name = "grid_north_pole_latitude",
             type = "NC_FLOAT",
             ucp$grid$pollat)
  att.put.nc(nc, variable = "rotated_pole", name = "grid_north_pole_longitude",
             type = "NC_FLOAT",
             ucp$grid$pollon)

  nc_def_var(nc, varname = "FR_URB", dimensions = c("rlon", "rlat"),
             standard_name = "fraction_urban",
             long_name = "fraction of urban surfaces",
             units = "1")

  nc_def_var(nc, varname = "FR_UCLASS", dimensions = c("rlon", "rlat", "uclass"),
             standard_name = "fraction_urban_classes",
             long_name = "fraction of urban classes",
             units = "1")

  nc_def_var(nc, varname = "FR_UDIR", dimensions = c("rlon", "rlat", "udir", "uclass"),
             standard_name = "fraction_street_direction",
             long_name = "fraction of street directions",
             units = "1")

  nc_def_var(nc, varname = "FR_ROOF", dimensions = c("rlon", "rlat", "uheight1", "udir", "uclass"),
             standard_name = "building_height_fraction",
             long_name = "building height fraction",
             units = "1")

  nc_def_var(nc, varname = "W_STREET", dimensions = c("rlon", "rlat", "udir", "uclass"),
             standard_name = "street_width",
             long_name = "street width",
             units = "m")

  nc_def_var(nc, varname = "W_BUILD", dimensions = c("rlon", "rlat", "udir", "uclass"),
             standard_name = "building_width",
             long_name = "building width",
             units = "m")

  att.put.nc(nc, variable = "NC_GLOBAL", name = "dcepucp",
             type = "NC_CHAR",
             paste("dcepucp R package version", utils::packageVersion("dcepucp"),
                   "(https://github.com/sebschub/dcepucp/)")
             )
  att.put.nc(nc, variable = "NC_GLOBAL", name = "creation_date",
             type = "NC_CHAR",
             as.character(Sys.time())
             )

  # data
  var.put.nc(nc, variable = "rlon",     data = rotated_longitude(ucp$grid))
  var.put.nc(nc, variable = "rlat",     data = rotated_latitude(ucp$grid))
  var.put.nc(nc, variable = "udir",     data = ucp$grid$angle_udir)
  var.put.nc(nc, variable = "uheight1", data = ucp$grid$hhl_uhl)

  var.put.nc(nc, variable = "lon", data = longitude(ucp$grid))
  var.put.nc(nc, variable = "lat", data = latitude(ucp$grid))

  var.put.nc(nc, variable = "FR_URB",    data = ucp$fr_urb)
  var.put.nc(nc, variable = "FR_UCLASS", data = ucp$fr_uclass)
  var.put.nc(nc, variable = "FR_UDIR",   data = ucp$fr_udir)
  var.put.nc(nc, variable = "FR_ROOF",   data = ucp$fr_roof)

  var.put.nc(nc, variable = "W_BUILD",  data = ucp$w_build)
  var.put.nc(nc, variable = "W_STREET", data = ucp$w_street)

  close.nc(nc)
}

data-raw/read_save_berlin_data.R
library(RNetCDF)
library(devtools)
library(dcepucp)

nc <- open.nc("berlin0.009.nc")
berlin_fr_roof <- var.get.nc(nc, "FR_ROOF", collapse = FALSE)
berlin_fr_uclass <- var.get.nc(nc, "FR_UCLASS", collapse = FALSE)
berlin_fr_udir <- var.get.nc(nc, "FR_UDIR", collapse = FALSE)
berlin_fr_urb <- var.get.nc(nc, "FR_URB", collapse = FALSE)
berlin_w_build <- var.get.nc(nc, "W_BUILD", collapse = FALSE)
berlin_w_street <- var.get.nc(nc, "W_STREET", collapse = FALSE)

berlin_rlon <- var.get.nc(nc, "rlon", collapse = FALSE)
berlin_rlat <- var.get.nc(nc, "rlat", collapse = FALSE)
berlin_uheight1 <- var.get.nc(nc, "uheight1", collapse = FALSE)
berlin_udir <- var.get.nc(nc, "udir", collapse = FALSE)

berlin_uclass <- dim.inq.nc(nc, "uclass")
close.nc(nc)

# change convention from angle 0 south to angle 0 north
swap_udir <- function(field) {
  temp <- field
  if (length(dim(field)) == 4) {
    field[ , , 1, ] <- temp[ , , 3, ]
    field[ , , 3, ] <- temp[ , , 1, ]
  } else if (length(dim(field)) == 5) {
    field[ , , , 1, ] <- temp[ , , , 3, ]
    field[ , , , 3, ] <- temp[ , , , 1, ]
  } else {
    stop("Wrong dimensions of field")
  }
  field
}

berlin_fr_roof  <- swap_udir(berlin_fr_roof)
berlin_fr_udir  <- swap_udir(berlin_fr_udir)
berlin_w_build  <- swap_udir(berlin_w_build)
berlin_w_street <- swap_udir(berlin_w_street)

berlin_grid <- ugrid(pollat = 37.483, pollon = -166.6,
                     startlat_tot = round(berlin_rlat[1], 4),
                     startlon_tot = round(berlin_rlon[1], 4),
                     ie_tot = length(berlin_rlon), je_tot = length(berlin_rlat),
                     dlat = round(berlin_rlat[2] - berlin_rlat[1], 4),
                     dlon = round(berlin_rlon[2] - berlin_rlon[1], 4),
                     ke_uhl = length(berlin_uheight1) - 1,
                     hhl_uhl = as.vector(berlin_uheight1),
                     n_uclass = berlin_uclass$length,
                     n_udir = length(berlin_udir),
                     angle_udir = as.vector(berlin_udir)
                     )

use_data(berlin_fr_roof, berlin_fr_uclass, berlin_fr_udir, berlin_fr_urb,
         berlin_w_build, berlin_w_street, berlin_grid,
         overwrite = TRUE)


R/data.R
#' Heights distribution of buildings in Berlin
#'
#' Height distribution of buildings in Berlin. The grid is defined in
#' \link{berlin_grid}.
#'
#' @format An array with dimensions "rlon", "rlat", "uheight1", "udir" and
#' "uclass" (see \link{ugrid}). In this example, only one urban class is considered.
#'
#' @source From \href{http://doi.org/10.1127/0941-2948/2013/0393}{Schubert and
#'   Grossman-Clarke 2013}.
"berlin_fr_roof"

#' Fraction of urban classes in Berlin
#'
#' Fraction of different urban classes in Berlin. The grid is defined in
#' \link{berlin_grid}. Since only one urban class is considered in this example,
#' the value is 1 where urban values are defined.
#'
#' @format An array with dimensions "rlon", "rlat"  and "uclass" (see
#'   \link{ugrid}). In this example, only one urban class is considered.
#'
#' @source From \href{http://doi.org/10.1127/0941-2948/2013/0393}{Schubert and
#'   Grossman-Clarke 2013}.
"berlin_fr_uclass"

#' Fraction of street directions in Berlin
#'
#' Fractions of different street directions in Berlin. The grid is defined in
#' \link{berlin_grid}.
#'
#' @format An array with dimensions "rlon", "rlat", "udir" and "uclass" (see
#'   \link{ugrid}). In this example, only one urban class is considered.
#'
#' @source From \href{http://doi.org/10.1127/0941-2948/2013/0393}{Schubert and
#'   Grossman-Clarke 2013}.
"berlin_fr_udir"

#' Urban fraction in Berlin
#'
#' Urban fraction of a grid cell in Berlin. The grid is defined in
#' \link{berlin_grid}.
#'
#' @format An array with dimensions "rlon" and "rlat" (see \link{ugrid}).
#'
#' @source From \href{http://doi.org/10.1127/0941-2948/2013/0393}{Schubert and
#'   Grossman-Clarke 2013}.
"berlin_fr_urb"

#' Building width in Berlin
#'
#' Building width in Berlin. The grid is defined in \link{berlin_grid}.
#'
#' @format An array with dimensions "rlon", "rlat", "udir" and "uclass" (see
#'   \link{ugrid}). In this example, only one urban class is considered.
#'
#' @source From \href{http://doi.org/10.1127/0941-2948/2013/0393}{Schubert and
#'   Grossman-Clarke 2013}.
"berlin_w_build"

#' Street width in Berlin
#'
#' Street width in Berlin. The grid is defined in \link{berlin_grid}.
#'
#' @format An array with dimensions "rlon", "rlat", "udir" and "uclass" (see
#'   \link{ugrid}). In this example, only one urban class is considered.
#'
#' @source From \href{http://doi.org/10.1127/0941-2948/2013/0393}{Schubert and
#'   Grossman-Clarke 2013}.
"berlin_w_street"

#' Used grid for the parameters of Berlin
#'
#' Urban grid used with urban Berlin files.
#'
#' @format Object of class \code{ugrid} (see \link{ugrid}).
#'
#' @source From \href{http://doi.org/10.1127/0941-2948/2013/0393}{Schubert and
#'   Grossman-Clarke 2013}.
"berlin_grid"

#' Default LCZ parameter
#'
#' Default Local Climate Zones parameter.
#'
#' @format Object of class \code{lcz}.
"lcz_default"